<a href="https://colab.research.google.com/github/bosco-morales/Deep-Learning-Practice/blob/main/CAP4936_Building_and_Enhancing_Binary_Classification_Models.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Bosco Moises Morales

Programming Assignment: Building and Enhancing Binary Classification Models

CAP4936 - Special Topics in Data Analytics

Summer 2024

## About


The dataset **AIDS Virus Infection Prediction** classifies the patients whether they are infected or not with HIV AIDS.

The dataset contains healthcare statistics and categorical information about patients who have been diagnosed with AIDS. This dataset was initially published in 1996.

[Click here to view dataset from Kaggle](https://www.kaggle.com/datasets/aadarshvelu/aids-virus-infection-prediction)

## Import Libraries

In [1]:
from pandas import read_csv
from sklearn.preprocessing import LabelEncoder

In [2]:
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

## Load Data

In [3]:
url = 'https://raw.githubusercontent.com/bosco-morales/Deep-Learning-Practice/main/Assignment%20%232%20Binary%20Classification/AIDS_Classification.csv'
data = read_csv(url, header=None)

In [4]:
data.head()

0    1    2        3     4     5      6       7       8    9   ...    13  \
0  time  trt  age     wtkg  hemo  homo  drugs  karnof  oprior  z30  ...  str2   
1   948    2   48  89.8128     0     0      0     100       0    0  ...     0   
2  1002    3   61  49.4424     0     0      0      90       0    1  ...     1   
3   961    3   45   88.452     0     1      1      90       0    1  ...     1   
4  1166    3   47  85.2768     0     1      0     100       0    1  ...     1   

      14       15     16      17    18     19    20     21        22  
0  strat  symptom  treat  offtrt  cd40  cd420  cd80  cd820  infected  
1      1        0      1       0   422    477   566    324         0  
2      3        0      1       0   162    218   392    564         1  
3      3        0      1       1   326    274  2063   1893         0  
4      3        0      1       0   287    394  1590    966         0  

[5 rows x 23 columns]

In [5]:
#Remove row 0 that includes headers
data.drop(index=0, inplace=True)

## Data Preprocessing

In [6]:
# Split data into input (X) and output (y)
X, y = data.values[:, :-1], data.values[:, -1]
X = X.astype('float32')

# Convert class labels to binary
y = LabelEncoder().fit_transform(y)

## Model Definition & Compilation

In [7]:
# Define the MLP model
model = Sequential()
model.add(Dense(10, activation='relu', input_shape=(X.shape[1],)))
model.add(Dense(1, activation='sigmoid'))

# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

## Model Training
        

In [8]:
# Fit the model to the data
model.fit(X, y, epochs=50, batch_size=32, verbose=0)

## Evaluation and Prediction

In [ ]:
# Evaluate the model
loss, accuracy = model.evaluate(X, y, verbose=0)
print(f"Model Accuracy: {accuracy:.3f}")

# Test a prediction
sample = numbers = numbers = [1002, 3, 61, 49.4424, 0, 0, 0, 90, 0, 1, 895, 0, 0, 1, 3, 0, 1, 0, 162, 218, 392, 564]
prediction = model.predict([sample])
print(f"Predicted Class: {round(prediction[0][0])}")

Model Accuracy: 0.848
1/1 [==============================] - 0s 29ms/step
Predicted Class: 0


## Comments

The model was evaluated with a record with class = 1 and the model predicted 0.

The model accuracy is 0.85